File: 15. Largest Rectangle in Histogram problem.cpp
/*
 Task: Largest Rectangle in Histogram problem
*/

#include <bits/stdc++.h>
using namespace std;

int largestRectangleArea(const vector<int>& heights)
{
    int n = heights.size();
    stack<int> st;
    int maxA = 0;
    for (int i = 0; i <= n; ++i)
    {
        int h = (i == n) ? 0 : heights[i];
        while (!st.empty() && h < heights[st.top()])
        {
            int height = heights[st.top()];
            st.pop();
            int width = st.empty() ? i : i - st.top() - 1;
            maxA = max(maxA, height * width);
        }
        st.push(i);
    }
    return maxA;
}

int main()
{
    vector<int> h = {2, 1, 5, 6, 2, 3};
    cout << largestRectangleArea(h) << "\n";
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)

File: 15. Largest Rectangle in Histogram problem.py
"""
Task: Largest Rectangle in Histogram problem

Time Complexity: O(n)
Space Complexity: O(n)
"""

def largest_rectangle_area(heights):
    n = len(heights)
    st = []
    maxA = 0
    for i in range(n+1):
        h = 0 if i==n else heights[i]
        while st and h < heights[st[-1]]:
            height = heights[st.pop()]
            width = i if not st else i - st[-1] - 1
            maxA = max(maxA, height*width)
        st.append(i)
    return maxA

if __name__ == '__main__':
    print(largest_rectangle_area([2,1,5,6,2,3]))
